fix(snap): allow disabling of secret-store via config hook

tonyespy opened the pull request:
This fix adds a new snap configure option 'security-secret-store'
that controls how secrets are stored. By default if a profile's
configuration file requires secrets, secret-store will be used.
If 'security-secret-store' is set to 'false', then insecure
secrets will be used instead. This aligns the snap with the same
capabilty available in the main edgexfoundry snap and the standard
docker-compose files.

Fixes: 150

Signed-off-by: Tony Espy <espy@canonical.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [NA] Tests for the changes have been added (for bug fixes / features)
- [NA] Docs have been added / updated (for bug fixes / features)

<!-- If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/.github/CONTRIBUTING.md -->

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->

Issue Number: 150

## What is the new behavior?
Secret store usage can now be disabled via a snap config option.


## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## Are there any new imports or modules? If so, what are they used for and why?
NA

## Are there any specific instructions or things that should be known prior to reviewing?
See issue

## Test Instructions
Install the edgexfoundy snap from latest/stable:

`$ sudo snap install edgexfoundry
`

Build edgex-app-service-configurable from this branch:

```
$ sudo snap install snapcraft --classic
$ snapcraft --use-lxd
$ sudo snap install edgex-app-service-configurable*amd64.snap --dangerous
```

Set the profile to mqtt-export (which uses secrets by default):

`$ sudo snap set edgex-app-service-configurable profile=mqtt-export
`

In a separate terminal start journalctl to follow system log messages:

`$ sudo journalctl -f -o cat
`

Start edgex-app-service-configurable and see that it fails to start.

`$ sudo snap start --enable edgex-app-service-configurable.app-service-configurable 
`

Stop it, set security-secret-store=off, and start it. This time it should come up with no errors logged.

```
$ sudo snap stop edgex-app-service-configurable.app-service-configurable
$ sudo snap set edgex-app-service-configurable security-secret-store=off
$ sudo snap start --enable edgex-app-service-configurable.app-service-configurable 
```

Verify that it up using snap services:

```
$ snap services edgex-app-service-configurable.app-service-configurable 
Service                                                  Startup   Current  Notes
edgex-app-service-configurable.app-service-configurable  enabled  active   -
```

lenny-intel commented on snap/README.md:
````diff
@@ -89,3 +89,31 @@ or
+$ sed -i -e 's@/vault/config/assets/resp-init.json@/var/snap/edgex-app-service-configurable/current/config/res/mqtt-export/secrets-token.json@' ./configuration.toml
+```
+
+**Note** -- these configuration changes need to be made *before* the service is started for the first time. Otherwise, the changes need to be made directly in Consul, and the service restarted.
````
One can also manually remove the the configuration from Consul and restart the service, which I recommend over making the changes in Consul

tonyespy replied:
That's true, but this note was about making secure secrets work, not disabling them. The latter is what this PR is all about (i.e. disabling secure secrets via the configure hook) so no Consul changes are required. I could add something about this earlier in the page, but I think it might confuse the average user.

siggiskulason replied:
Following the steps above, I get journal messages saying the following. Could be my setup:

 http: TLS handshake error from 127.0.0.1:47914: remote error: tls: bad certificate

lenny-intel replied:
> but this note was about making secure secrets work, not disabling them

My comment is aligned with `making secure secrets work`. Has nothing to do with disabling. Pointing out it is easier to simply remove the configuration and let the service re-seed it, than make the edits in Consul.

tonyespy replied:
@lenny-intel I took your suggestion and re-worded this note to recommend deleting the configuration and starting over. Just re-pushed, so marking this resolved.

tonyespy commented:
Testing the changes works as expected, so I've added test instructions to the PR description and removed the draft status.

siggiskulason commented on snap/README.md:
````diff
@@ -89,3 +89,31 @@ or
+Here's an example of how to use a configuration/profile which includes secrets:
+
+```
+$ sudo snap install edgex-app-service-configurable --channel=latest/candidate/alpha
````
Is latest/candidate/alpha the correct channel?

tonyespy replied:
nice catch!

siggiskulason commented on snap/README.md:
````diff
@@ -89,3 +89,31 @@ or
+```
+$ sudo snap install edgex-app-service-configurable --channel=latest/candidate/alpha
+$ sudo snap set profile=mqtt
+$ cd /var/snap/edgex-app-service-configurable/current/config/res/mqtt-profile
````
Should this be mqtt-export? I don't see a mqtt-profile directory in this location

tonyespy replied:
This was a cut & paste which should have been reviewed/edited. Thanks for pointing these out.

siggiskulason commented on snap/README.md:
````diff
@@ -89,3 +89,31 @@ or
+$ cd /var/snap/edgex-app-service-configurable/current/config/res/mqtt-profile
+$ sudo cp /var/snap/edgexfoundry/secrets/ca/ca.pem .
+$ sudo cp /var/snap/edgexfoundry/secrets/edgex-application-service/secrets-token.json .
+```
````
is /current missing from the path? i.e. should this be /var/snap/edgexfoundry/current/secrets/

siggiskulason commented on snap/README.md:
````diff
@@ -89,3 +89,31 @@ or
+```
+
+Next the profile's configuration.toml file needs to be updated to reference the paths in ```$SNAP_DATA``` of the copied files. The two config fields that need updated are ```  RootCaCertPath``` and ```TokenFile```. The following shows how this can be done via the sed command-line tool, however the file can also be easily updated via your favorite editor.
+
````
So the configuration.toml file in /var/snap/edgex-app-service-configurable/current/config/res/mqtt-export ?

siggiskulason commented:
Looks good to me. I tried the test instructions above and only comment there is that 
sudo snap install edgex-app-service-configurable*amd64.snap
needs to be
sudo snap install --dangerous edgex-app-service-configurable*amd64.snap

Other than that see inline comments. I tried testing with a certificate as well as per the README but for TLS handshake errors. Didn't investigate that further so could just be my setup

tonyespy commented:
I just force-pushed an update addressing most of the comments concerning `snap/README.md`.

@siggiskulason re: your test results, I'm trying to reproduce now to verify my test instructions.

tonyespy commented:
@siggiskulason OK, I think I've addressed all of your comments in `snap/README.md`, and I also re-tested the manual steps to enable vault support (ie. the last section of the README which provides instructions to copy `ca.pem` and `secrets-token.json`) and verified that they now work.

Please re-review at your convenience.

siggiskulason reviewed (approved):
LGTM

lenny-intel reviewed (approved):
LGTM
